### ISSUE TITLE
Texture compute support

### ISSUE THREAD

**Issue by habemus-papadum** [tsherif/picogl.js#138]
These features would be useful in developing/debugging `texture compute` (ala tensor.js): 

(Quick and dirty -- will return later to clarify.)

* It's an odd use case, but it would be useful to specify a mip level when attaching a texture to frame buffer (in the  readframebuffer role) for readPixels (Currently hard coded to mip level: 0) (see https://stackoverflow.com/questions/50979565/summing-the-values-in-a-webgl2-r32f-texture-by-generating-a-mipmap)

* Utils read back data in textures (I have some code that does this in part but for not all formats, also there are some browser bugs to note: /https://github.com/KhronosGroup/WebGL/issues/2747 )

*  I think these lines need to be added to texture.data()
````
this.gl.pixelStorei(GL.UNPACK_FLIP_Y_WEBGL, this.flipY);
        this.gl.pixelStorei(GL.UNPACK_PREMULTIPLY_ALPHA_WEBGL, this.premultiplyAlpha);
````

Sorry for the sloppy entry -- just making a quick note before I forget.  Will cleanup later/hopefully add some pull request.

**Comment by tsherif** [comment]
All seems pretty reasonable. For the second item would it just be a matter of changing `App.readPixel` to allow specifying the width and height?

**Comment by habemus-papadum** [comment]
@tsherif Not sure if github pings you when entries are edited, so adding this comment just in case.

**Comment by tsherif** [comment]
Thanks for the thorough treatment! This sounds very interesting. I'll read in more detail later and make comments later, but one note is that a core part of how I've developed PicoGL is to go "example first", i.e. choose an algorithm I want to implement and build out the API to support it (I should probably document this somewhere). There are two reasons for this:
- Ensures APIs are built in a way that "feels good" in a real-world use case
- The examples act as integration tests for the library
So what I'd suggest is start by building out that reaction diffusion example and adding the API features required to support it.

**Comment by tsherif** [comment]
Oh and to answer your questions about FBOs. I believe all attachment to an FBO have to be the same size, but I'll double check that...

**Comment by habemus-papadum** [comment]
Example first sounds good!

**Comment by habemus-papadum** [comment]
So I spent some time investigating some details regarding textures, reading and writing them , how the format parameters work etc.  Ultimately I should write a one or two page tutorial regarding the details that are important for machine learning and physics simulations, but for now I'm going to place bits and pieces here.

* in the call to `texSubImage2d`  (for example, similar points apply to 3D, etc)  there are three format  parameters: `InternalFormat`, `Format`, `Type` in addition to the data parameter -- which for now we will always assume is a `TypedArray`.  
* in `readPixels` there is only `Format`, `Type`, plus a `TypedArray` data parameter
* `Format` and `Type`   (I believe) are only relevant to specifying how data should be marshalled into and out of the `TypeArray`.  Once the marshalling is done, they are no longer relevant.  I'm ~80% confident that's how this works.
* `InternalFormat` will govern how much GPU memory is used, whether shaders should use `isampler2D`, `usampler2D`, `sampler2D`, etc
* With regards to marshalling, the basic questions are which `TypedArray` to use, how many elements does it need to have, and in come cases how to encode and decode data into the proper bit format. 
* The mapping from `Type` to `TypedArray` is specified in  section 3.7.6 Texture Objects of the WebGL2 Spec:

````
TypedArray  WebGL Type 
----------  ----------
Int8Array	BYTE
Uint8Array	UNSIGNED_BYTE
Int16Array	SHORT
Uint16Array	UNSIGNED_SHORT
Uint16Array	UNSIGNED_SHORT_5_6_5
Uint16Array	UNSIGNED_SHORT_5_5_5_1
Uint16Array	UNSIGNED_SHORT_4_4_4_4
Int32Array	INT
Uint32Array	UNSIGNED_INT
Uint32Array	UNSIGNED_INT_5_9_9_9_REV
Uint32Array	UNSIGNED_INT_2_10_10_10_REV
Uint32Array	UNSIGNED_INT_10F_11F_11F_REV
Uint32Array	UNSIGNED_INT_24_8
Uint16Array	HALF_FLOAT
Float32Array	FLOAT
````
* The number of elements in the array and precisely which elements will be read and which are skipped are controlled by `FORMAT`, `TYPE`, the texture type (2d, 3d, etc), the texture dimensions, and the `pixelStorei` parameters
* Again, the `pixelStorei` parameters are related only to marshalling data only (again, so I believe)

So one basic question that needs to be decided is that for the purposes of picogl, should a `Texture` keep info related to marshalling data as part of its state (as it does now), or should they be passed each time when reading and writing (and for convenience in the constructor).  

My guess is that there aren't many use cases where data is being marshalled into and out of texture in multiple formats  but since picogl does a good job of keeping possibilities as open as possible I figured I would at least raise the question.  Probably a reasonable compromise is to always remember the last used marshalling parameters, and use them by default, if they are not specified.
